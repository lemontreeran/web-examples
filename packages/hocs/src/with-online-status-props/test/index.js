import React from 'react'
import { mount } from 'enzyme'

const Target = () => null
const dummyMapStatusToProps = () => {}

describe('withOnlineStatusProps', () => {
  describe('Online Status API is supported', () => {
    let origOnlineStatus = null
    let withOnlineStatusProps = null

    beforeAll(() => {
      origOnlineStatus = global.navigator.onLine
    })

    beforeEach(() => {
      jest.resetModules()

      Object.defineProperty(global.navigator, 'onLine', {
        get: () => true,
        configurable: true
      })
      withOnlineStatusProps = require('../src/').default
    })

    afterAll(() => {
      Object.defineProperty(global.navigator, 'onLine', {
        get: () => origOnlineStatus,
        configurable: true
      })
    })

    it('should set initial state', () => {
      const EnhancedTarget = withOnlineStatusProps(
        ({ isOnline, isOffline }) => ({ isOnline, isOffline })
      )(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      expect(wrapper.find(Target)).toMatchSnapshot()
    })

    it('should handle online status change', () => {
      const EnhancedTarget = withOnlineStatusProps(
        ({ isOnline, isOffline }) => ({ isOnline, isOffline })
      )(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )

      global.dispatchEvent(new CustomEvent('offline'))
      wrapper.update()
      expect(wrapper.find(Target)).toMatchSnapshot()
      global.dispatchEvent(new CustomEvent('online'))
      wrapper.update()
      expect(wrapper.find(Target)).toMatchSnapshot()
    })

    it('should remove event listener on unmount', () => {
      const EnhancedTarget = withOnlineStatusProps(
        ({ isOnline, isOffline }) => ({ isOnline, isOffline })
      )(Target)
      const wrapper = mount(
        <EnhancedTarget />
      )
      const instance = wrapper.instance()

      wrapper.unmount()
      global.document.dispatchEvent(new CustomEvent('offline'))
      // TODO: mount again and check?
      // https://github.com/airbnb/enzyme/pull/969
      expect(instance.state).toMatchSnapshot()
    })

    describe('display name', () => {
      let origNodeEnv = null

      beforeAll(() => {
        origNodeEnv = process.env.NODE_ENV
      })

      afterAll(() => {
        process.env.NODE_ENV = origNodeEnv
      })

      it('should wrap display name in non-production env', () => {
        process.env.NODE_ENV = 'test'

        const EnhancedTarget = withOnlineStatusProps(dummyMapStatusToProps)(Target)
        const wrapper = mount(
          <EnhancedTarget />
        )

        expect(wrapper).toMatchSnapshot()
      })

      it('should not wrap display name in production env', () => {
        process.env.NODE_ENV = 'production'

        const EnhancedTarget = withOnlineStatusProps(dummyMapStatusToProps)(Target)
        const wrapper = mount(
          <EnhancedTarget />
        )

        expect(wrapper).toMatchSnapshot()
      })
    })
  })

  describe('Online Status API is not supported', () => {
    let origOnlineStatus = null
    let withOnlineStatusProps = null

    beforeAll(() => {
      origOnlineStatus = global.navigator.onLine
    })

    beforeEach(() => {
      jest.resetModules()

      Object.defineProperty(global.navigator, 'onLine', {
        get: () => {},
        configurable: true
      })
      withOnlineStatusProps = require('../src/').default
    })

    afterAll(() => {
      Object.defineProperty(global.navigator, 'onLine', {
        get: () => origOnlineStatus,
        configurable: true
      })
    })

    it('should just pass Target component through', () => {
      const EnhancedTarget = withOnlineStatusProps(dummyMapStatusToProps)(Target)
      const wrapper = mount(
        <EnhancedTarget a={1} b={2} c={3} />
      )

      expect(wrapper).toMatchSnapshot()
    })
  })
})
