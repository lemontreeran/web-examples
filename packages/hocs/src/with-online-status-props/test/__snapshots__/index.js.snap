// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withOnlineStatusProps Online Status API is not supported should just pass Target component through 1`] = `
<Target
  a={1}
  b={2}
  c={3}
/>
`;

exports[`withOnlineStatusProps Online Status API is supported display name should not wrap display name in production env 1`] = `
<WithOnlineStatusProps>
  <Target />
</WithOnlineStatusProps>
`;

exports[`withOnlineStatusProps Online Status API is supported display name should wrap display name in non-production env 1`] = `
<withOnlineStatusProps(Target)>
  <Target />
</withOnlineStatusProps(Target)>
`;

exports[`withOnlineStatusProps Online Status API is supported should handle online status change 1`] = `
<Target
  isOffline={true}
  isOnline={false}
/>
`;

exports[`withOnlineStatusProps Online Status API is supported should handle online status change 2`] = `
<Target
  isOffline={false}
  isOnline={true}
/>
`;

exports[`withOnlineStatusProps Online Status API is supported should remove event listener on unmount 1`] = `
Object {
  "isOffline": false,
  "isOnline": true,
}
`;

exports[`withOnlineStatusProps Online Status API is supported should set initial state 1`] = `
<Target
  isOffline={false}
  isOnline={true}
/>
`;
